br/pkg/restore/misc.go

```diff
@@ -44,6 +53,182 @@ const (
+	}
+	if !bytes.Equal(markerFile.checksumLogRestoreTableIDsMarkerFile(), markerFile.Checksum) {
+		return 0, 0, nil, errors.Errorf(
+			"checksum mismatch (calculated checksum is %s but the recorded checksum is %s), the log restore table IDs blocklist file may be corrupted",
+			base64.StdEncoding.EncodeToString(markerFile.checksumLogRestoreTableIDsMarkerFile()),
+			base64.StdEncoding.EncodeToString(markerFile.Checksum),
+		)
+	}
+	return markerFile.RestoreCommitTs, markerFile.SnapshotBackupTs, markerFile.TableIds, nil
+}
+
+func fastWalkLogRestoreTableIDsMarkerFile(
+	ctx context.Context,
+	s storage.ExternalStorage,
+	filterOutFn func(restoreCommitTs, snapshotBackupTs uint64) bool,
+	executionFn func(ctx context.Context, filename string, restoreCommitTs uint64, tableIds []int64) error,
+) error {
+	filenames := make([]string, 0)
+	if err := s.WalkDir(ctx, &storage.WalkOption{SubDir: logRestoreTableIDMarkerFilePrefix}, func(path string, _ int64) error {
+		restoreCommitTs, snapshotBackupTs, parsed := parseLogRestoreTableIDsMarkerFileName(path)
+		if parsed {
+			if filterOutFn(restoreCommitTs, snapshotBackupTs) {
+				return nil
+			}
+		}
+		filenames = append(filenames, path)
+		return nil
+	}); err != nil {
+		return errors.Trace(err)
+	}
+	workerpool := tidbutil.NewWorkerPool(8, "walk dir log restore table IDs blocklist files")
+	eg, ectx := errgroup.WithContext(ctx)
+	for _, filename := range filenames {
+		if ectx.Err() != nil {
+			break
+		}
+		workerpool.ApplyOnErrorGroup(eg, func() error {
+			data, err := s.ReadFile(ectx, filename)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			restoreCommitTs, snapshotBackupTs, tableIds, err := unmarshalLogRestoreTableIDsMarkerFile(data)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if filterOutFn(restoreCommitTs, snapshotBackupTs) {
+				return nil
+			}
+			err = executionFn(ectx, filename, restoreCommitTs, tableIds)
+			return errors.Trace(err)
+		})
+	}
+	return errors.Trace(eg.Wait())
+}
+
+// CheckTableTrackerContainsTableIDsFromMarkerFiles checks whether pitr id tracker contains the filtered table IDs from blocklist file.
+func CheckTableTrackerContainsTableIDsFromBlocklistFiles(
+	ctx context.Context,
+	s storage.ExternalStorage,
+	tracker *utils.PiTRIdTracker,
+	startTs, restoredTs uint64,
+) error {
+	err := fastWalkLogRestoreTableIDsMarkerFile(ctx, s, func(restoreCommitTs, snapshotBackupTs uint64) bool {
+		return startTs >= restoreCommitTs || restoredTs <= snapshotBackupTs
+	}, func(_ context.Context, _ string, restoreCommitTs uint64, tableIds []int64) error {
+		for _, tableId := range tableIds {
+			if tracker.ContainsTableId(tableId) || tracker.ContainsPartitionId(tableId) {
+				return errors.Errorf(
+					"cannot restore the table(Id=%d) because it is restored(at %d) before snapshot backup(at %d). "+
+						"Please respecify the filter that does not contain the table or replace with a newer snapshot backup.",
+					tableId, restoreCommitTs, startTs)
```

Comment:
It's better to record more table info(like table name),  do we have chance to make this log more easy to debug?